interop-cats/jvm/src/main/scala/scalaz/zio/interop/catsjvm.scala

```diff
@@ -30,54 +30,56 @@ abstract class CatsPlatform extends CatsInstances {
+    override def shift: TaskR[R, Unit] =
+      ZIO.yieldNow
 
-    override def evalOn[A](ec: ExecutionContext)(fa: IO[E, A]): IO[E, A] =
+    override def evalOn[A](ec: ExecutionContext)(fa: TaskR[R, A]): TaskR[R, A] =
       fa.on(ec)
   }
 
-  implicit def ioTimer[R <: Clock, E]: effect.Timer[ZIO[R, E, ?]] = new effect.Timer[ZIO[R, E, ?]] {
-    override def clock: cats.effect.Clock[ZIO[R, E, ?]] = new effect.Clock[ZIO[R, E, ?]] {
-      override def monotonic(unit: TimeUnit): ZIO[R, E, Long] =
+  implicit def ioTimer[R <: Clock]: effect.Timer[TaskR[R, ?]] = new effect.Timer[TaskR[R, ?]] {
+    override def clock: cats.effect.Clock[TaskR[R, ?]] = new effect.Clock[TaskR[R, ?]] {
+      override def monotonic(unit: TimeUnit): TaskR[R, Long] =
         zioClock.nanoTime.map(unit.convert(_, NANOSECONDS))
 
-      override def realTime(unit: TimeUnit): ZIO[R, E, Long] =
+      override def realTime(unit: TimeUnit): TaskR[R, Long] =
         zioClock.currentTime(unit)
     }
 
-    override def sleep(duration: FiniteDuration): ZIO[R, E, Unit] =
+    override def sleep(duration: FiniteDuration): TaskR[R, Unit] =
       zioClock.sleep(scalaz.zio.duration.Duration.fromNanos(duration.toNanos))
   }
 
-  implicit val taskEffectInstances: effect.ConcurrentEffect[Task] with SemigroupK[Task] =
-    new CatsConcurrentEffect
+  implicit def taskEffectInstances[R]: effect.ConcurrentEffect[TaskR[R, ?]] with SemigroupK[TaskR[R, ?]] =
+    new CatsConcurrentEffect[R]
 
   implicit val taskParallelInstance: Parallel[Task, Util.Par] =
     parallelInstance(taskEffectInstances)
 }
 
 sealed abstract class CatsInstances1 extends CatsInstances2 {
-  implicit def ioMonoidInstances[E: Monoid]: MonadError[IO[E, ?], E] with Bifunctor[IO] with Alternative[IO[E, ?]] =
-    new CatsAlternative[E] with CatsBifunctor
+  implicit def ioMonoidInstances[R, E: Monoid]
+    : MonadError[ZIO[R, E, ?], E] with Bifunctor[ZIO[R, ?, ?]] with Alternative[ZIO[R, E, ?]] =
+    new CatsAlternative[R, E] with CatsBifunctor[R]
 
-  implicit def parallelInstance[E](implicit M: Monad[IO[E, ?]]): Parallel[IO[E, ?], ParIO[E, ?]] =
-    new CatsParallel[E](M)
+  implicit def parallelInstance[R, E](implicit M: Monad[ZIO[R, E, ?]]): Parallel[ZIO[R, E, ?], ParIO[E, ?]] =
+    new CatsParallel[R, E](M)
 }
 
 sealed abstract class CatsInstances2 {
-  implicit def ioInstances[E]: MonadError[IO[E, ?], E] with Bifunctor[IO] with SemigroupK[IO[E, ?]] =
-    new CatsMonadError[E] with CatsSemigroupK[E] with CatsBifunctor
+  implicit def ioInstances[R]
+    : MonadError[TaskR[R, ?], Throwable] with Bifunctor[ZIO[R, ?, ?]] with SemigroupK[TaskR[R, ?]] =
+    new CatsMonadError[R, Throwable] with CatsSemigroupK[R, Throwable] with CatsBifunctor[R]
 }
 
-private class CatsConcurrentEffect extends CatsConcurrent with effect.ConcurrentEffect[Task] {
+private class CatsConcurrentEffect[R] extends CatsConcurrent[R] with effect.ConcurrentEffect[TaskR[R, ?]] {
```

Comment:
For this type class, I think we will have to hard code `R` to be some known set of environmental effects (say, all the effects supported by `DefaultRuntime`), so that we can provide them before we (for example) convert `toIO`.

Basically if you look at Cats `ReaderT[R, F, ?]`, we can support all the cats effects type classes that it supports, without hard coding `R`; then for the rest, we need to pick an `R`. Either that or use implicits to "inject" an `R`.

Let's have @kaishh take a look too since this could hugely improve usability of `R` for cats-effect users...

Reply:
Yeah, saying `[R >: DefaultRuntime]` would allow an implementation of `toIO` - but it wouldn't be all that useful for someone trying to use `R` as Reader since they would be limited to only `R >: DefaultRuntime <: Any` - @gvolpe please correct me if you think otherwise. It would also invite incoherence, because the `DefaultRuntime` provided will NOT be the one created by user, but just `CatsConcurrentEffect`'s own `this`!

The most principled solution I see would unfortunately _decrease_ usability for cats-effect users :)
- Provide instances up to Concurrent for any `R`
- Provide an instance for ConcurrentEffect with Effect for `R <: Runtime[R]` where implicit R is required and STOP extending DefaultRuntime since we can use the executor from implicit instead of creating a global

Another is to simply provide *Effect only for `R=Any`, this means no implicit rituals and no incoherence in environment – at the cost of not being able to access CatsConcurrentEffect.this - but I don't think that's an awful loss.

Reply:
@jdegoes @kaishh thanks for looking into it. 

> Provide instances up to Concurrent for any R

I think this makes sense and this is the way it's done in `cats-effect`. Instances for `Kleisli` can only be implemented up to `Concurrent` but somehow I thought having the Reader capabilities built in would allow to go further.

> Another, more usable solution, is to simply provide *Effect only for R=Any

This seems to be the best solution given the first point :)

Will give it a spin later today.

Reply:
> Another, more usable solution, is to simply provide *Effect only for R=Any

Unless I'm missing something, I'm not a fan of this approach as it doesn't _increase_ usability.

I really think that for up to `Concurrent`, we should be maximally polymorphic and support _any_ `R`, because it has no impact on cats-effect users: if they're not using `R` (`R =:= Any`), then they can still benefit from the polymorphic instances; and if they _are_ using `R`, they'll be grateful that "just works" out of the box with everything up to `Concurrent` — and the only reason it doesn't work with classes beyond that is because of the run methods and `toIO` methods that will be going away in Cats Effect 2.0.

The "global injector" is interesting. Right now you have zero control over thread pool in Cats. You're stuck with the default one. No chance to use, for example, Scala's global execution context or an app-specific thread pool.

If we stop extending `DefaultRuntime`, then `Concurrent` and beyond could require an implicit `Runtime[R]`, which allows a user to do two things:

 - Use any `R` for all the type classes
 - Configure the platform (thread pool, fatal errors, error reporter)

This comes at cost to adding an implicit, but we could make it easy, e.g.:

```scala
import scalaz.zio.cats.implicits.DefaultRuntime
```

How does this sound?

Reply:
@jdegoes I think @kaishh meant `R=Any` for `Effect` and `ConcurrentEffect`. All the instances up to `Concurrent` are now fully polymorphic on `R`.

With regards to the runtime, I'll leave that decision to you both that are more familiar with the design :)

Reply:
All right, great. In that case, for `Concurrent` and beyond, my vote would be having those instances take an implicit `Runtime[R]` so we can make them all polymorphic in `R`, and then adding a global implicit so users can import to "unsafe run" stuff that just uses `R = Console with Random with System with Blocking with Scheduler`.